NorthstarDLL/client/debugoverlay.cpp

```diff
@@ -119,21 +212,136 @@ void, __fastcall, (OverlayBase_t * pOverlay))
 		}
 	}
 	break;
+	case OVERLAY_TRIANGLE:
+	{
+		OverlayTriangle_t* pTriangle = static_cast<OverlayTriangle_t*>(pOverlay);
+		RenderTriangle(
+			pTriangle->p1,
+			pTriangle->p2,
+			pTriangle->p3,
+			Color(pTriangle->r, pTriangle->g, pTriangle->b, pTriangle->a),
+			pTriangle->noDepthTest);
 	}
-	LeaveCriticalSection((LPCRITICAL_SECTION)((char*)sEngineModule + 0x10DB0A38));
+	break;
+	case OVERLAY_SWEPT_BOX:
+	{
+		OverlaySweptBox_t* pBox = static_cast<OverlaySweptBox_t*>(pOverlay);
+		RenderWireframeSweptBox(
+			pBox->start, pBox->end, pBox->angles, pBox->mins, pBox->maxs, Color(pBox->r, pBox->g, pBox->b, pBox->a), false);
+	}
+	break;
+	case OVERLAY_SPHERE:
+	{
+		OverlaySphere_t* pSphere = static_cast<OverlaySphere_t*>(pOverlay);
+		RenderSphere(
+			pSphere->vOrigin,
+			pSphere->flRadius,
+			pSphere->nTheta,
+			pSphere->nPhi,
+			Color(pSphere->r, pSphere->g, pSphere->b, pSphere->a),
+			false);
+	}
+	break;
+	default:
+	{
+		spdlog::warn("Unimplemented overlay type {}", pOverlay->m_Type);
+	}
+	break;
+	}
+
+	LeaveCriticalSection(s_OverlayMutex);
+}
+
+// clang-format off
+AUTOHOOK(DrawAllOverlays, engine.dll + 0xAB780, 
+void, __fastcall, (bool bRender))
+// clang-format on
+{
+	EnterCriticalSection(s_OverlayMutex);
+
+	OverlayBase_t* pCurrOverlay = *s_pOverlays; // rbx
+	OverlayBase_t* pPrevOverlay = nullptr; // rsi
+	OverlayBase_t* pNextOverlay = nullptr; // rdi
+
+	int m_nCreationTick; // eax
+	bool bShouldDraw; // zf
+	int m_pUnk; // eax
+
+	while (pCurrOverlay)
+	{
+		if (OverlayBase_t__IsDead(pCurrOverlay))
+		{
+			if (pPrevOverlay)
+			{
+				pPrevOverlay->m_pNextOverlay = pCurrOverlay->m_pNextOverlay;
+			}
+			else
+			{
+				*s_pOverlays = pCurrOverlay->m_pNextOverlay;
+			}
+
+			pNextOverlay = pCurrOverlay->m_pNextOverlay;
+			OverlayBase_t__DestroyOverlay(pCurrOverlay);
+			pCurrOverlay = pNextOverlay;
+		}
+		else
+		{
+			if (pCurrOverlay->m_nCreationTick == -1)
+			{
+				m_pUnk = pCurrOverlay->m_pUnk;
+
+				if (m_pUnk == -1)
+				{
+					bShouldDraw = true;
+				}
+				else
+				{
+					bShouldDraw = m_pUnk == *g_nOverlayTickCount;
+				}
+			}
+			else
+			{
+				bShouldDraw = pCurrOverlay->m_nCreationTick == *g_nRenderTickCount;
+			}
+
+			if (bShouldDraw && bRender && (Cvar_enable_debug_overlays->GetBool() || pCurrOverlay->m_Type == OVERLAY_SMARTAMMO))
```

Comment:
Running this shouldn't impact performance, but its pontless to do so if bRender is false.

Maybe this should be in an `if (bRender) {}`? might hurt readability though.

Reply:
> Running this shouldn't impact performance, but its pontless to do so if bRender is false.

C++ has short circuiting, so this really doesn't matter tbh

Reply:
eh, sure.
I was mostly thinking about not running code thats not needed, but its so little code t hat any CPU capable of running TF|2 should have this done in a few cycles.

Reply:
If you mean running the majority of the function, it's needed so that expired debug draws get cleared from the list

Reply:
Nah, I just mean the part directly above the if condition